### PR TITLE
 adding a request to the Apply step logic.

### PR DESCRIPTION
**Pull request by bxy4543**
adding a request to the Apply step logic.
confirm actions when deleting and scaling down, do we need confirmation to add nodes?

### CODE REVIEW

File: common/structs.go [hunk 1]
```diff
@@ -26,3 +26,5 @@ type RunArgs struct {
 	Provider   string
 	CustomEnv  []string
 }
+
+var ForceDelete bool
```
Comment (kakaZhou719) [inline common/structs.go:30]: its not a good way to use global variable. we can add a flag on Applier

File: apply/applydriver/local.go [hunk 2]
```diff
@@ -149,7 +149,7 @@ func (c *Applier) scaleCluster(mj, md, nj, nd []string) error {
 		return nil
 	}
 
-	logger.Info("Start to scale this cluster")
+	logger.Info("current cluster: master %s, worker %s", c.ClusterCurrent.GetMasterIPList(), c.ClusterCurrent.GetNodeIPList())
```
Comment (fanux) [inline apply/applydriver/local.go:152]: Start to scale this cluster, current cluster...

File: apply/applydriver/local.go [hunk 3]
```diff
@@ -252,7 +252,14 @@ func (c *Applier) initCluster() error {
 }
 
 func (c *Applier) deleteCluster() error {
-	logger.Info("Start to delete current cluster")
+	logger.Info("Start to delete cluster: master %s, node %s", c.ClusterDesired.GetMasterIPList(), c.ClusterDesired.GetNodeIPList())
```
Comment (fanux) [inline apply/applydriver/local.go:255]: I think, add it before call `runtime.Delete` is more security.

File: pkg/runtime/runtime.go [hunk 4]
```diff
@@ -56,25 +62,58 @@ func (k *KubeadmRuntime) Upgrade() error {
...
+	if len(newNodesIPList) != 0 {
+		logger.Info("%s will be added as worker", newNodesIPList)
+	}
 	return k.joinNodes(newNodesIPList)
 }
 
 func (k *KubeadmRuntime) DeleteMasters(mastersIPList []string) error {
+	if len(mastersIPList) != 0 {
+		logger.Info("master %s will be deleted", mastersIPList)
+		if err := k.confirmDeleteNodes(); err != nil {
```
Comment (fanux) [inline pkg/runtime/runtime.go:89]: We also need to support force delete, like in e2e test

File: apply/applydriver/local.go [hunk 5]
```diff
@@ -149,7 +149,7 @@ func (c *Applier) scaleCluster(mj, md, nj, nd []string) error {
 		return nil
 	}
 
-	logger.Info("Start to scale this cluster")
+	logger.Info("Start to scale this cluster, current cluster: master %s, worker %s", c.ClusterCurrent.GetMasterIPList(), c.ClusterCurrent.GetNodeIPList())
```
Comment (fanux) [inline apply/applydriver/local.go:152]: What if it has 100 nodes, the iplist will very long

Comment (bxy4543) [reply apply/applydriver/local.go:152]: > What if it has 100 nodes, the iplist will very long

How about changing the log level of the current cluster to DEBUG mode?

Comment (fanux) [reply apply/applydriver/local.go:152]: logger.Info("Start to scale this cluster")
logger.Debug("current cluster:.....")

File: pkg/runtime/runtime.go [hunk 6]
```diff
@@ -56,25 +62,58 @@ func (k *KubeadmRuntime) Upgrade() error {
...
+	}
 	return k.deleteNodes(nodesIPList)
 }
 
+func (k *KubeadmRuntime) confirmDeleteNodes() error {
+	if !ForceDelete {
+		if pass, err := utils.ConfirmOperation("Are you sure to delete these nodes? "); err != nil {
+			return err
+		} else if !pass {
+			os.Exit(0)
```
Comment (allencloud) [inline pkg/runtime/runtime.go:111]: just return nil ?

**Review by fanux — APPROVED** [review]

**Pull request merged**